crates/state/src/write_log.rs

```diff
@@ -168,6 +170,39 @@ impl WriteLog {
         }
     }
 
+    /// Read a non-temporary value at the given key and return the value and the
+    /// gas cost, returns [`None`] if the key is not present in the write
+    /// log
+    pub fn read_persistent(
+        &self,
+        key: &storage::Key,
+    ) -> (Option<&StorageModification>, u64) {
```

Comment:
I think in general this all looks great, but this function return type is a bit strange as in type in can return `StorageModification::Temp` so it's a bit weird that we then have to match on it after calling `read_persistent`. It will be a bit more verbose here but perhaps we should add e.g. another `enum PersistentModification`